### PR TITLE
Add functions for slicing distributed arrays locally

### PR DESCRIPTION
**Pull request by NanmiaoWu**
This PR:

1. Renames `slice_row_d` and `slice_column_d` to denote the slice operation that works on distributed arrays with index on the whole  array.

2. Adds functions for assigning 1d vector to distributed array with index working locally. This will return the distributed array (with some rows being modified) still with annotation.

3. Adds tests for extracting 1d vector from distributed array with local index by using the functions for non-distributed arrays. This will return some 1d vectors without annotation.

### CODE REVIEW

File: src/execution_tree/primitives/slice.cpp [hunk 1]
```diff
@@ -275,16 +280,85 @@ namespace phylanx { namespace execution_tree
                     name, codename));
         }
 
-        if (data.has_annotation() || value.has_annotation())
+//        if (data.has_annotation() || value.has_annotation())
```
Comment (taless474) [inline src/execution_tree/primitives/slice.cpp:283]: I guess we can remove these commented lines

File: src/plugins/matrixops/slicing_operation.cpp [hunk 2]
```diff
@@ -74,27 +74,71 @@ namespace phylanx {namespace execution_tree {    namespace primitives
...
 
+    ///////////////////////////////////////////////////////////////////////////
+    namespace detail
+    {
+        slicing_operation::slice_mode slicing_operation_mode(
+            std::string const& name)
+        {
+            slicing_operation::slice_mode result
+                = slicing_operation::local_mode;
+            if ((name.find("slice_row_d") != std::string::npos) ||
```
Comment (taless474) [inline src/plugins/matrixops/slicing_operation.cpp:123]: There is something that I am not sure I understand. Does `slice_d` work on its own?

File: src/execution_tree/primitives/slice.cpp (hunk 1 above)
Comment (NanmiaoWu) [reply src/execution_tree/primitives/slice.cpp:283]: Yes. Will delete it.

File: src/plugins/matrixops/slicing_operation.cpp (hunk 2 above)
Comment (NanmiaoWu) [reply src/plugins/matrixops/slicing_operation.cpp:123]: No. I think I need to delete slice_d here.

**Pull request merged**